File: leetcode/186-reverse-string-ii/main.go
package main

import "fmt"

/*
	1st approach
	1. reverse the whole string
	2. iterate forward, reverse a word in the back if encounter a space/end of a string
	Time	O(n) 2n->n
	Space	O(1) in-place
	1096ms beats 100%
*/

func reverseWords(str []byte) {
	reverseInplace(str, 0, len(str)-1)
	start := 0
	for i := 0; i < len(str); i++ {
		if str[i] == ' ' {
			reverseInplace(str, start, i-1)
			start = i + 1
		} else if i == len(str)-1 {
			reverseInplace(str, start, i)
			start = i + 1
		}
	}
}

func reverseInplace(str []byte, start, end int) {
	for start < end {
		str[start], str[end] = str[end], str[start]
		start++
		end--
	}
}

func main() {
	a := []byte{'a', 'b', ' ', 'c', 'd'}
	reverseWords(a)
	fmt.Println(a)
}

File: leetcode/186-reverse-string-ii/main.py
"""
    1st approach
	1. reverse the whole string
	2. iterate forward, reverse a word in the back if encounter a space/end of a string
    
	Time	O(n) 2n->n
	Space	O(1) in-place
	264 ms, faster than 52.17%
"""


class Solution:
    def reverseWords(self, s: List[str]) -> None:
        n = len(s)
        self.reverse(s, 0, n-1)
        slow = 0
        for i in range(n):
            if i+1 == n:
                self.reverse(s, slow, i)
            elif s[i] == ' ':
                self.reverse(s, slow, i-1)
                slow = i + 1

    def reverse(self, s, i, j):
        while i < j:
            s[i], s[j] = s[j], s[i]
            i += 1
            j -= 1
